### ISSUE TITLE
SPAdes: processing collection problem

### ISSUE THREAD

**Issue by gallardoalba** [galaxyproject/tools-iuc#4658]
Currently, datasets from collections are processed together. This behavior should be modified in order to assemble each dataset from the collections independently.

**Comment by bernt-matthias** [comment]
I guess there are usecases for batch and joint processing. So you know how other assembly tools are implemented in this regard? Would be best to have both supported by a conditional or a new  param type.

**Comment by gallardoalba** [comment]
Yes, I think you are right; I'll include a new param for including both alternatives.

**Comment by JayanthiGangiredla** [comment]
Hi, I still see spades is processing datasets from collections together instead of running them individually.

**Comment by gallardoalba** [comment]
Hi @JayanthiGangiredla, yes, I didn't include the modifications yet. I'll try to work on it the next week.

**Comment by bernt-matthias** [comment]
The solution is to restructure the collections. If your input is a paired list, then it needs to be restructured as list of paired lists (where the inner lists each contains one pair). The restructuring can be done using the `Apply rules` tool.

**Comment by JayanthiGangiredla** [comment]
Is there a tutorial on this? i am unable to make spades run on multiple samples at a time. Same for Flye assembller as well.

**Comment by bernt-matthias** [comment]
> Is there a tutorial on this?

Not to my knowledge unfortunatelly.

Have you tried the `Apply rules` tool? It should work as follows:

Select a collection

Edit rules

- You should see a table with a single column
- Column -> Basename of Path or URL -> Choose column `A`
- Rules -> Add / Modify ... -> Add Definition  -> List Identifiers A
-                           Assign another column -> List Identifiers B
- Apply
- Save

Run tool

Then you have a nested collection that you can feed to the tool.

**Comment by JayanthiGangiredla** [comment]
Thank you so much! I have used buildlist instead on an interlaced reads, now spades running them as a nested list.  This is the case with Flye assembler as well. Now this is resolved! Thank youfor being very helpful when we needed!  :)

**Comment by jorondo1** [comment]
@bernt-matthias I have tried your suggestion
> Have you tried the `Apply rules` tool? It should work as follows:
> 
> Select a collection
> ...
> Then you have a nested collection that you can feed to the tool.

Passing this nested collection works, but the only way metaspades to recognize this is if I select the interlaced reads mode. That's fine but why isn't passing a pair possible? Or do I have to interlace my pair and then nest the collection?

**Comment by bernt-matthias** [comment]
A collection of collection of pairs should work.

1st level contains the samples
2nd level is a list containing 1 pair.

**Comment by jorondo1** [comment]
Yes that's exactly what I have, yet I can only find it if I put metaspades in interlaced modes, which fails when there is an uneven number of reads. 

<img width="1355" height="518" alt="Image" src="https://github.com/user-attachments/assets/3c76aa95-0bf6-441d-a1b6-88f42848a5ad" />

Does the list look right? The input list to which I applied the procedure you suggested above contained "a list with 5 fastqsanger.gz pairs" and the output "a list with 5 fastqsanger.gz lists". That collection can't be selected with `Paired-end : list of dataset pairs` mode.

Edit: Should I try interlacing my reads ? I've never had to do that with metaspades from the command line (as best as I can remember).

**Comment by bernt-matthias** [comment]
It seems that you have a list of list, where the inner list has length 2. But you would need a list of list of pairs.

If you are on a public Galaxy you may also share your history.

**Comment by jorondo1** [comment]
Here : https://usegalaxy.org/u/jonathan.rondeau.leclaire/h/full-wms-pipeline

**Comment by bernt-matthias** [comment]
Here is how I would do this: 

https://github.com/user-attachments/assets/0f6c5dcc-b0b6-46a3-b7e3-7dbca5f49965

The resulting history is here: 

https://usegalaxy.org/u/m.bernt/h/copy-of-full-wms-pipeline

**Comment by jorondo1** [comment]
That works great! Thanks so much. This is slightly different from the procedure you gave above, there was no mention of column C (then again I wasn't sure what I was doing when I followed it!)

Would it make sense to have a small tool that's a "paired-end list nester" or something like that? I feel like this is normally the default usage for assemblers, as co-assembly is normally used in special cases where samples come from the same same individual, same sampling site ,etc. Also co-assembling dozens of sample uses an insane amount of memory. 

Maybe the metaspades tool inside galaxy could simply do this list transformation in the background except if someone explicitly asks for co-assembly, with a switch like the Megahit tool. 

I have no idea how the backend of galaxy works, but this would be my suggestion!

thanks again for your help
